go/runtime/proxy.go

```diff
@@ -0,0 +1,434 @@
+	if req.Open == nil {
+		return fmt.Errorf("missing open message")
+	}
+	if err := req.Open.ShardId.Validate(); err != nil {
+		return fmt.Errorf("invalid shard id: %w", err)
+	}
+	if req.Open.TargetPort == 0 {
+		return fmt.Errorf("missing target port")
+	}
+	if req.Open.TargetPort > 65535 {
+		return fmt.Errorf("target port '%d' out of range", req.Open.TargetPort)
+	}
+	if req.Open.ClientAddr == "" {
+		return fmt.Errorf("missing client addr")
+	}
+	if len(req.Data) > 0 {
+		return fmt.Errorf("first proxy message cannot have both Open and Data")
+	}
+
+	return nil
+}
+
+func copyResponses(streaming pf.NetworkProxy_ProxyServer, client pf.NetworkProxy_ProxyClient, shard string, port uint32, outboundBytes *uint64) error {
+	var counter = proxyConnBytesOutboundCounter.WithLabelValues(shard, strconv.Itoa(int(port)))
+	for {
+		var resp, err = client.Recv()
+		if err != nil {
+			return err
+		}
+
+		// Validate the response by ensuring it has only Data
+		if resp.OpenResponse != nil {
+			return fmt.Errorf("data response contained OpenResponse")
+		}
+		if len(resp.Data) == 0 {
+			return fmt.Errorf("response did not contain data")
+		}
+		if err = streaming.Send(resp); err != nil {
+			return err
+		}
+		counter.Add(float64(len(resp.Data)))
+		atomic.AddUint64(outboundBytes, uint64(len(resp.Data)))
+	}
+}
+
+func copyRequests(streaming pf.NetworkProxy_ProxyServer, client pf.NetworkProxy_ProxyClient, shard string, port uint32, inboundBytes *uint64) error {
+	var counter = proxyConnBytesInboundCounter.WithLabelValues(shard, strconv.Itoa(int(port)))
+	defer client.CloseSend()
+	for {
+		var req, err = streaming.Recv()
+		if err != nil {
+			return err
+		}
+		// Validate the request by ensuring it has only Data
+		if req.Open != nil {
+			return fmt.Errorf("data message contained Open")
+		}
+		if len(req.Data) == 0 {
+			return fmt.Errorf("request did not contain data")
+		}
+		if err = client.Send(req); err != nil {
+			return err
+		}
+		counter.Add(float64(len(req.Data)))
+		atomic.AddUint64(inboundBytes, uint64(len(req.Data)))
+	}
+}
+
+type runningContainer struct {
+	instanceVersion int
+	ports           map[uint16]bool
+	connection      *grpc.ClientConn
+	logger          ops.Publisher
+}
+
+type networkConfigHandle struct {
+	server *ProxyServer
+	// container lifecycles aren't always as straight forward as you'd hope for.
+	// It can happen that a container exits, and `Unexpose` is called, before
+	// `Expose` is called. This incrementing number is used to ensure correctness
+	// in scenarios where things may be called out of order.
+	instanceVersion int
+	shardID         pc.ShardID
+	ports           map[uint16]*labels.PortConfig
+}
+
+func (h *networkConfigHandle) Expose(connection *grpc.ClientConn, logger ops.Publisher) {
+	if h.instanceVersion > 0 {
+		panic("expose called more than once")
+	}
+
+	h.server.mu.Lock()
+	defer h.server.mu.Unlock()
+
+	// copy the ports map to ensure it doesn't get modified while we're using it.
```

Comment:
Ya know, that's a good point. On reflection, I don't think this copy is necessary at all. They come from the `taskTerm`, which intentionally never modifies them after initialization. I just updated it to use the same ports map instead of creating a new one.